riscv-c-api.md

```diff
@@ -213,3 +213,66 @@ long __riscv_clmul (long a, long b); // clmul rd, rs1, rs2
 #include <riscv_vector.h> // make RISC-V vector intrinsics available
 vint8m1_t __riscv_vadd_vv_i8m1(vint8m1_t vs2, vint8m1_t vs1, size_t vl); // vadd.vv vd, vs2, vs1
 ```
+
+## Constraints on Operands of Inline Assembly Statements
+
+This section lists operand constraints that can be used with inline assembly
+statements, including both RISC-V specific and common operand constraints.
+
+
+| Constraint         |                                    | Note        |
+| ------------------ | ---------------------------------- | ----------- |
+| m                  | An address that is held in a general-purpose register with offset.      |      |
+| A                  | An address that is held in a general-purpose register.      |      |
+| r                  | General purpose register                      |      |
+| f                  | Floating-point register                       |      |
+| i                  | Immediate integer operand                     |      |
+| I                  | 12-bit signed immediate integer operand      |      |
+| K                  | 5-bit unsigned immediate integer operand     |      |
+| J                  | Zero integer immediate operand                |      |
+
+NOTE: Immediate value must be a compile-time constant.
+
+### The Difference Between `m` and `A` Constraints
+
+The difference between `m` and `A` is whether the operand can have an offset;
+some instructions in RISC-V do not allow an offset for the address operand,
+such as atomic or vector load/store instructions.
+
+The following example demonstrates the difference; it is trying
+to load value from `foo[10]` and using `m` and `A` to pass that address.
+
+```c
+int *foo;
+void bar() {
+ int x;
+ __asm__ volatile ("lw %0, %1" : "=r"(x) : "m" (foo[10]));
+ __asm__ volatile ("lw %0, %1" : "=r"(x) : "A" (foo[10]));
+}
+```
+
+Then we compile with GCC with `-O` option:
+
+```shell
+$ riscv64-unknown-elf-gcc x.c -o - -O -S
+...
+bar:
+       lui    a5,%hi(foo)
+       ld     a5,%lo(foo)(a5)
+ #APP
```

Comment:
What does "APP" stand for?

Reply:
APP is comment markers from Clang's inline assembly handling code...

Reply:
Seems like it has some function on gnu Assembler Used to turning preprocessing on and off:

https://sourceware.org/binutils/docs/as/Preprocessing.html
```
If the first line of an input file is #NO_APP or if you use the ‘-f’ option, whitespace and comments are not removed from the input file. Within an input file, you can ask for whitespace and comment removal in specific portions of the file by putting a line that says #APP before the text that may contain whitespace or comments, and putting a line that says #NO_APP after this text. This feature is mainly intended to support asm statements in compilers whose output is otherwise free of comments and whitespace.
```

And from GCC doc: https://gcc.gnu.org/onlinedocs/gccint/File-Framework.html
```
Macro: ASM_APP_ON
A C string constant for text to be output before each asm statement or group of consecutive ones. Normally this is "#APP", which is a comment that has no effect on most assemblers but tells the GNU assembler that it must check the lines that follow for all valid assembler constructs.

Macro: ASM_APP_OFF
A C string constant for text to be output after each asm statement or group of consecutive ones. Normally this is "#NO_APP", which tells the GNU assembler to resume making the time-saving assumptions that are valid for ordinary compiler output.
```


But I guess that is just a marker for clang/llvm.

BTW, that's the first time I figure out what APP/NO_APP means.